# CMPS 140

# Assignment 3

**Due July 14, 2019 (Bastille day!) 11:59**

In [1]:
NAME = "Derrick DeBose"

## Problem 1

Consider the following constraint satisfaction problem. A linear graph has nodes of the following colors:

- Red
- Yellow
- Green
- Blue
- Violet

Each node has a domain of {1, 2, ..., 9}.

Each node type as the following constraints on its value:

- Red - No contraints
- Yellow - The rightmost digit of of the product of all its neighbors
- Green - The rightmost digit of the sum of all its neighbors 
- Blue - The leftmost digit of the sum of all its neighbors
- Violet - The leftmost digit of the product of all of its neightbors

As a reminder here is the pseudo code for the Min-Conflicts search algorithm:

![minconflicts](https://docs.google.com/drawings/d/e/2PACX-1vTIdOyAKDEoK6evNWQBkx9X5kl2I7GLaUkE9TdFDRqyyNFiHeFDrA-Sm7sLob2wMSzoBk_cliRhs8PY/pub?w=927&amp;h=474)

**Notes:**

- It's possible that you won't converge to a solution in a single run. Try a few runs to see if you get to a solution.
- The example is to show you what a problem looks like, we will test/grade your program on different examples


Complete the function *solve_csp* defined below. You may find some helper functions useful.

In [2]:
#return true if no constraint conflict otherwise returns false
def no_conflict(nodes,arcs,node_values):
    node_length = len(nodes)
    arc_length = len(arcs)
    #need to make sure each node has no conflicts
    for i in range(0,node_length):
        if nodes[i] == 'R':
            #Red has no constraints that need satisfying
            continue
        if nodes[i] == 'Y':
            #Yellow - The rightmost digit of of the product of all its neighbors
            product = 1
            for j in range(0,arc_length):
                if(arcs[j][0]==i  or arcs[j][1] ==i):
                    #determines the node values of the current nodes neighbors
                    if arcs[j][0]==i:
                        product = product * node_values[arcs[j][1]]
                    if arcs[j][1]==i:
                        product = product * node_values[arcs[j][0]]
            product = product%10
            #if its neighbors' values total to product there is no conflict
            if node_values[i]!=product:
                return False
        if nodes[i] == 'G':
            #Green - The rightmost digit of the sum of all its neighbors
            sum_ = 0
            for j in range(0,arc_length):
                if(arcs[j][0]==i  or arcs[j][1] ==i):
                    #determines the node values of the current nodes neighbors
                    if arcs[j][0]==i:
                        sum_ = sum_ + node_values[arcs[j][1]]
                    if arcs[j][1]==i:
                        sum_ = sum_ + node_values[arcs[j][0]]
            sum_ = sum_%10
            #if its neighbors' values total to sum_ there is no conflict
            if node_values[i]!=sum_:
                return False
        if nodes[i] == 'B':
            #Blue - The leftmost digit of the sum of all its neighbors
            sum_ = 0
            for j in range(0,arc_length):
                if(arcs[j][0]==i  or arcs[j][1] ==i):
                    #determines the node values of the current nodes neighbors
                    if arcs[j][0]==i:
                        sum_ = sum_ + node_values[arcs[j][1]]
                    if arcs[j][1]==i:
                        sum_ = sum_ + node_values[arcs[j][0]]
            while sum_ >= 10:
                sum_ = sum_//10
            #if its neighbors' values total to sum_ there is no conflict
            if node_values[i] != sum_:
                return False
        if nodes[i] == 'V':
            #Violet - The leftmost digit of the product of all of its neightbors
            product = 1
            for j in range(0,arc_length):
                if(arcs[j][0]==i  or arcs[j][1] ==i):
                    #determines the node values of the current nodes neighbors
                    if arcs[j][0]==i:
                        product = product * node_values[arcs[j][1]]
                    if arcs[j][1]==i:
                        product = product * node_values[arcs[j][0]]
            while product >= 10:
                product = product//10
            #if its neighbors' values total to product there is no conflict
            if node_values[i] != product:
                return False
    #after checking all the nodes and none returned False then
    #it is true that there are no conflict
    return True



def solve_csp(nodes, arcs, max_steps):
    """
    This function solves the csp using the MinConflicts Search
    Algorithm.

    INPUTS:
    nodes:      a list of letters that indicates what type of node it is,
                the index of the node in the list indicates its id
                letters = {R, Y, G, B, V}
                
    arcs:       a list of tuples that contains two numbers, indicating the 
                IDS of the nodes the arc connects. 
                
    max_steps:  max number of steps to make before giving up

    RETURNS: a list of values for the solution to the CSP where the 
             index of the value corresponds the value for that
             given node.
    """
    from random import randrange
    
    
    for _ in range(max_steps):
        #we want to check 1000 different random numbers
        node_values = [randrange(1,10), randrange(1,10), randrange(1,10), randrange(1,10), randrange(1,10)]
        for _ in range(max_steps):
            #as we try to fix the conflicts. we want to check if there is no conflicts at any state
            if no_conflict(nodes,arcs,node_values):
                return node_values
            else:
                #there is a conflict
                arc_length = len(arcs)
                for i in range(0,len(nodes)):
                    #as we try to fix the conflicts. we want to check if there is no conflicts at any state
                    if no_conflict(nodes,arcs,node_values):
                        return node_values
                    
                    else:
                        if nodes[i] == 'R':
                            if no_conflict(nodes,arcs,node_values):
                                return node_values
                            #change value of red if theres a conflict
                            if node_values[i] == 9:
                                node_values[i] = 1
                            else:
                                node_values[i] = node_values[i] + 1
                        if nodes[i] == 'Y':
                            #Yellow - The rightmost digit of of the product of all its neighbors
                            product = 1
                            for j in range(0,arc_length):
                                if(arcs[j][0]==i  or arcs[j][1] ==i):
                                    #determines the node values of the current nodes neighbors
                                    if arcs[j][0]==i:
                                        product = product * node_values[arcs[j][1]]
                                    if arcs[j][1]==i:
                                        product = product * node_values[arcs[j][0]]

                            product = product%10
                            #checker to see if current color has no conflicts
                            #change value if theres a conflict
                            if node_values[i] != product:
                                if node_values[i] == 9:
                                    node_values[i] = 1
                                else:
                                    node_values[i] = node_values[i] +1
                        if nodes[i] == 'G':
                            #Green - The rightmost digit of the sum of all its neighbors
                            sum_ = 0
                            for j in range(0,arc_length):
                                if(arcs[j][0]==i  or arcs[j][1] ==i):
                                    #determines the node values of the current nodes neighbors
                                    if arcs[j][0]==i:
                                        sum_ = sum_ + node_values[arcs[j][1]]
                                    if arcs[j][1]==i:
                                        sum_ = sum_ + node_values[arcs[j][0]]
                            sum_ = sum_%10
                            #checker to see if current color has no conflicts
                            #change value if theres a conflict
                            if node_values[i] != sum_:
                                if node_values[i] == 9:
                                    node_values[i] = 1
                                else:
                                    node_values[i] = node_values[i] +1
                        if nodes[i] == 'B':
                            #Blue - The leftmost digit of the sum of all its neighbors
                            sum_ = 0
                            for j in range(0,arc_length):
                                if(arcs[j][0]==i  or arcs[j][1] ==i):
                                    #determines the node values of the current nodes neighbors
                                    if arcs[j][0]==i:
                                        sum_ = sum_ + node_values[arcs[j][1]]
                                    if arcs[j][1]==i:
                                        sum_ = sum_ + node_values[arcs[j][0]]
                            while sum_ >= 10:
                                sum_ = sum_//10
                            #checker to see if current color has no conflicts
                            #change value if theres a conflict
                            if node_values[i] != sum_:
                                if node_values[i] == 9:
                                    node_values[i] = 1
                                else:
                                    node_values[i] = node_values[i] +1
                        if nodes[i] == 'V':
                            #Violet - The leftmost digit of the product of all of its neightbors
                            product = 1
                            for j in range(0,arc_length):
                                if(arcs[j][0]==i  or arcs[j][1] ==i):
                                    #determines the node values of the current nodes neighbors
                                    if arcs[j][0]==i:
                                        product = product * node_values[arcs[j][1]]
                                    if arcs[j][1]==i:
                                        product = product * node_values[arcs[j][0]]
                            while product >= 10:
                                product = product//10
                            #checker to see if current color has no conflicts
                            #change value if theres a conflict
                            if node_values[i] != product:
                                if node_values[i] == 9:
                                    node_values[i] = 1
                                else:
                                    node_values[i] = node_values[i] +1
    #if we never found a solution just return the current node values
    return node_values


### Test Cases

Below we've included 4 test cases to help you debug your code. Your submitted code will be tested on different cases than these, but if your implementation of the above Min-Conflicts search algorithm pseudocode is able to solve these problems, you should be good to go.

In [3]:
# Test Case 1

nodes = 'YGVRB'
arcs = [(0,1), (0,2), (1,2), (1,3), (1,4), (2,3), (2,4)]
max_steps = 1000

for _ in range(max_steps):
    sol = solve_csp(nodes, arcs, max_steps)
    if sol != []:
        break
        
all_solutions = [[1, 1, 1, 7, 2],[2, 1, 2, 4, 3],[2, 6, 7, 6, 1],[2, 8, 9, 6, 1],
                 [3, 3, 1, 5, 4],[6, 2, 8, 7, 1],[6, 7, 8, 2, 1],[6, 9, 4, 8, 1]]

if sol == []:
    print('No solution')
else:
    if sol in all_solutions:
        print('Solution found:', sol)
    else:
        print('ERROR: False solution found:', sol)

Solution found: [2, 1, 2, 4, 3]


In [4]:
# Test Case 2

nodes = 'YVBGR'
arcs = [(0,1), (0,2), (1,3), (2,4)]
max_steps = 1000

for _ in range(max_steps):
    sol = solve_csp(nodes, arcs, max_steps)
    if sol != []:
        print(nodes)
        break
        
all_solutions = [[1, 1, 1, 1, 9], [1, 3, 7, 3, 6], [1, 7, 3, 7, 2], [1, 9, 9, 9, 8]]

if sol == []:
    print('No solution')
else:
    if sol in all_solutions:
        print('Solution found:', sol)
    else:
        print('ERROR: False solution found:', sol)

YVBGR
Solution found: [1, 1, 1, 1, 9]


In [5]:
# Test Case 3

nodes = 'VYGBR'
arcs = [(0,1), (1,2), (2,3), (3,4)]
max_steps = 1000

for _ in range(max_steps):
    sol = solve_csp(nodes, arcs, max_steps)
    if sol != []:
        print(nodes)
        break
        
all_solutions = [[2, 2, 1, 9, 8],[3, 3, 1, 8, 7],[4, 4, 1, 7, 6],[5, 5, 1, 6, 5],[5, 5, 3, 8, 5],
                 [6, 6, 1, 5, 4],[7, 7, 1, 4, 3],[8, 8, 1, 3, 2],[8, 8, 6, 8, 2],[9, 9, 1, 2, 1]]

if sol == []:
    print('No solution')
else:
    if sol in all_solutions:
        print('Solution found:', sol)
    else:
        print('ERROR: False solution found:', sol)

VYGBR
Solution found: [5, 5, 1, 6, 5]


In [6]:
# Test Case 4

nodes = 'YGVBR'
arcs = [(0,1), (0,2), (1,3), (2,3), (3,4), (1,2)]
max_steps = 1000

for _ in range(max_steps):
    sol = solve_csp(nodes, arcs, max_steps)
    if sol != []:
        print(nodes)
        break
        
all_solutions = [[4, 4, 1, 9, 4],[4, 7, 2, 1, 1],[4, 7, 2, 1, 2],[4, 7, 2, 1, 3],[4, 7, 2, 1, 4],
                 [4, 7, 2, 1, 5],[4, 7, 2, 1, 6],[4, 7, 2, 1, 7],[4, 7, 2, 1, 8],[4, 7, 2, 1, 9],
                 [4, 8, 3, 1, 1],[4, 8, 3, 1, 2],[4, 8, 3, 1, 3],[4, 8, 3, 1, 4],[4, 8, 3, 1, 5],
                 [4, 8, 3, 1, 6],[4, 8, 3, 1, 7],[4, 8, 3, 1, 8],[5, 1, 5, 1, 4],[5, 1, 5, 1, 5],
                 [5, 1, 5, 1, 6],[5, 1, 5, 1, 7],[5, 1, 5, 1, 8],[5, 1, 5, 1, 9]]

if sol == []:
    print('No solution')
else:
    if sol in all_solutions:
        print('Solution found:', sol)
    else:
        print('ERROR: False solution found:', sol)

YGVBR
Solution found: [4, 7, 2, 1, 1]


## Problem 2   -  the n rooks problem

Rooks can move any number of squares horizontally or vertically on a chess board. The n rooks problem is to arrange rooks on an $n\times n$ board in such a way that none of the rooks could bump into another by making any of its possible horizontal or vertical moves. 

For this problem, the variables are each column (labeled 0, 1, ... , $n-1$), the the domain consists of each possible row (also labeled 0, 1, ... ,$ n-1$). In each column we place a rook on row 0, row 1, ... , row $n-1$.

For example, if $n=2$ we have only two solutions to this problem:

    R @
    @ R 
    
or

    @ R
    R @ 
   
How many possible solutions are there to this CSP, in terms of n?  
Also, give a simple proof that your answer is correct.

### Answer: There are n! possible solutions to this CSP 
We can place down the first rook in any n number of columns and any n number of rows. Then when we place down the next rook there will now be n-1 rows and columns that rook can be placed in because 1 column and 1 row can no longer be selected because of the first placed down rook constraining the next rooks values. Until we get to the last rook that needs to be placed in and there is only one place that rook can be placed without it conflicting with the already placed down rooks.

### Python Library to solve CSPs

One useful Python module for solving these types of problems is called *constraint*. In the next cell you'll see how to load this module into a Jupyter notebook running in Colab. 

We'll use this module to solve the following simple CSP. Suppose our variables are x and y, and the values they are allowed to assume are numbers in the domain 

            {1, 2, ... , 100}, 

and with constraints be that $x^2=y$ and that $x$ is odd.

In [7]:
# The following line imports the constraint module into a Jupyter notebook running in Colab.
# (This only needs to be done once per session)
!pip install python-constraint

# Let's load all the functions available in this module
# (This only needs to be done once per session)
from constraint import *

# Now we initialize a new problem and add the two variables.
problem = Problem()
problem.addVariables(["x", "y"], list(range(1, 101)))

# So far there are no constraints.  If we solve the problem with no constraints, 
# then every possible combination of x and y values from 1 to 100 are valid solutions.
solutions = problem.getSolutions()

# There are 100*100 = 10000 solutions in there:
print( len(solutions) )

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Created wheel for python-constraint: filename=python_constraint-1.4.0-py2.py3-none-any.whl size=24081 sha256=b712d1b09448695b2825a5a0372134620b9aa67302b666f8665e59741dce6fbc
  Stored in directory: /root/.cache/pip/wheels/07/27/db/1222c80eb1e431f3d2199c12569cb1cac60f562a451fe30479
Successfully built python-constraint
10000


In [8]:
# Add the given constraints:
problem.addConstraint(lambda a, b: a**2 == b and (a % 2) == 1, ("x", "y"))

# There are only 5 solutions that satisfy those constraints:
solutions = problem.getSolutions()
print( len(solutions) )

5


In [9]:
# Here are all 5 solutions:
print(solutions)

[{'x': 9, 'y': 81}, {'x': 7, 'y': 49}, {'x': 5, 'y': 25}, {'x': 3, 'y': 9}, {'x': 1, 'y': 1}]


Notice that the solutions consist of a list of dictionaries, where each dictionary represents a solution. For example, the first solution is {'x': 9, 'y': 81}, since with $x=9$ and $y=81$ it's true that $x^2=y$.

### The n-rooks problem revisited

The folling code block shows a solution to this problem using the *constraint* module, with $n=8$.

In [10]:
# The n rooks problem

problem = Problem()
n = 8
cols = range(n)
rows = range(n)
problem.addVariables(cols, rows)
for c1 in cols:
    for c2 in cols:
        if c1 < c2:
            problem.addConstraint(lambda r1, r2: r1 != r2, (c1, c2))

solutions = problem.getSolutions()
len(solutions)

40320

## Problem 3  -  The n-queens problem

In chess, a queen can make all the moves of a rook, but also can move along diagonals. As before, our goal is to arrange $n$ queens on an $n\times n$ board so that none of their moves are impeded by the presense of another queen on any row, column, or diagonal.

Modify the $n$ rooks problem to solve the n queens problem. You only need one call to *problem.addConstraint*, as before, since you can use logical 'and's to include as many other constraints in the lambda function as you need.

In [11]:
# The n queens problem

problem = Problem()
n = 8
cols = range(n)
rows = range(n)
problem.addVariables(cols, rows)
for c1 in cols:
    for c2 in cols:
        if c1 < c2:
            # if you subtract the rows and columns of the 2 queens 
            # and they are equal to each other then the queens conflict on the diagonals
            problem.addConstraint(lambda r1, r2, a=c1, b=c2: r1 != r2 and abs(r1 - r2) !=abs(a - b), (c1, c2))
            
solutions = problem.getSolutions()
print(len(solutions))

# The n queens problem

problem = Problem()
n = 11
cols = range(n)
rows = range(n)
problem.addVariables(cols, rows)
for c1 in cols:
    for c2 in cols:
        if c1 < c2:
            # if you subtract the rows and columns of the 2 queens 
            # and they are equal to each other then the queens conflict on the diagonals
            problem.addConstraint(lambda r1, r2, a=c1, b=c2: r1 != r2 and abs(r1 - r2) != abs(a - b), (c1, c2))
            
solutions = problem.getSolutions()
print(len(solutions))

92
2680


## Problem 4  -  The n-queens problem

Note: You don't need to prove these, just report what you get from your above code.

(a) How many ways are there of arranging 8 queens on an 8x8 board so that none impede the others?

(b) How many ways are there arranging 11 queens on an 11x11 board so that none impede the others?

### Answer: 

#### (a) 92 
#### (b) 2680

